src/main/java/com/airtribe/newsaggregator/news/NewsService.java

```diff
@@ -0,0 +1,131 @@
+        User user = userRepository.findByEmail(userEmail)
+                .orElseThrow(() -> new UsernameNotFoundException("User not found"));
+        user.getReadArticles().add(articleId);
+        userRepository.save(user);
+    }
+
+    public Set<String> getReadArticles(String userEmail) {
+        User user = userRepository.findByEmail(userEmail)
+                .orElseThrow(() -> new UsernameNotFoundException("User not found"));
+        return user.getReadArticles();
+    }
+
+    @Transactional
+    public void markArticleAsFavorite(String userEmail, String articleId) {
+        User user = userRepository.findByEmail(userEmail)
+                .orElseThrow(() -> new UsernameNotFoundException("User not found"));
+        user.getFavoriteArticles().add(articleId);
+        userRepository.save(user);
+    }
+
+    public Set<String> getFavoriteArticles(String userEmail) {
+        User user = userRepository.findByEmail(userEmail)
+                .orElseThrow(() -> new UsernameNotFoundException("User not found"));
+        return user.getFavoriteArticles();
+    }
+
+    @Transactional
+    public NewsPreferencesResponse updateNewsPreferences(String userEmail, UpdateNewsPreferencesRequest updateRequest) {
+        User user = userRepository.findByEmail(userEmail)
+                .orElseThrow(() -> new UsernameNotFoundException("User not found"));
+        NewsPreference preferences = newsPreferenceRepository.findByUser(user);
+        if (preferences == null) {
+            preferences = new NewsPreference();
+            preferences.setUser(user);
+        }
+        preferences = newsPreferenceMapper.fromUpdateRequest(updateRequest);
+        preferences.setUser(user);
+        preferences = newsPreferenceRepository.save(preferences);
+        return newsPreferenceMapper.toDTO(preferences);
+    }
+
+    @Cacheable(value = "newsCache", key = "#userEmail")
+    public String fetchNews(String userEmail) {
+        User user = userRepository.findByEmail(userEmail)
+                .orElseThrow(() -> new UsernameNotFoundException("User not found"));
+        NewsPreference preferences = newsPreferenceRepository.findByUser(user);
+        String url = buildNewsApiUrl(preferences);
+        return restClient
+                .get()
+                .uri(url)
+                .retrieve()
+                .body(String.class);
+    }
+
+    @Scheduled(fixedRate = 3600000) // Update cache every hour
+    @CacheEvict(value = "newsCache", allEntries = true)
+    public void refreshNewsCache() {
+        userRepository.findAll().forEach(user -> {
+            String userEmail = user.getEmail();
+            fetchNews(userEmail); // This will repopulate the cache with updated data
+        });
+    }
+
+    @Cacheable(value = "newsSearchCache", key = "#keyword")
+    public String searchNewsByKeyword(String keyword) {
+        String url = String.format("https://gnews.io/api/v4/search?q=%s&apikey=%s", keyword, apiKey);
```

Comment:
`https://gnews.io/api/v4/search?q=%s&apikey=%s` this should be a config constant and not hardcoded